### PR TITLE
Add condition scope type to C#

### LINKED ISSUE #674: C# does not yet support the 'condition' scope type
It may be as simple as using the pattern matcher for 'conditional_expression'

### PR DESCRIPTION
**Pull request by AndrewDant**
Closes #674
## Checklist

- [ X] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)

### CODE REVIEW

**Review by auscompgeek — COMMENTED** [review]
Could you also add tests where the condition is:

- a simple value, e.g. `if (false) {}`
- includes logical not
- is a method call

**Comment by AndrewDant** [comment]
@auscompgeek I've implemented the first two tests you asked for. Can you give me an example of which you want for a method call? I don't think the condition type matches method call arguments in existing implementations.

**Comment by auscompgeek** [comment]
> I don't think the condition type matches method call arguments in existing implementations.

Huh, that sounds like a bug to me. Should fix that later if that's the case.

**Review by auscompgeek — APPROVED** [review]
LGTM

**Comment by pokey** [comment]
> @auscompgeek I've implemented the first two tests you asked for. Can you give me an example of which you want for a method call? I don't think the condition type matches method call arguments in existing implementations.

I would do `if (foo(bar)) {}` => `if () {}`. Does that not work?

File: src/languages/csharp.ts [hunk 1]
```diff
@@ -218,6 +220,10 @@ const nodeMatchers: Partial<Record<ScopeType, NodeMatcherAlternative>> = {
   ifStatement: "if_statement",
   class: "class_declaration",
   className: "class_declaration[name]",
+  condition: cascadingMatcher(
+    conditionMatcher("*[condition]"),
+    patternMatcher("~binary_expression.binary_expression!")
```
Comment (pokey) [inline src/languages/csharp.ts:225]: This seems odd. Wouldn't this match eg the right hand side of `a = b + c`?

**Review by pokey — CHANGES_REQUESTED** [review]

**Comment by AndrewDant** [comment]
> I would do `if (foo(bar)) {}` => `if () {}`. Does that not work?

Yes that does work and I've added a test for it. I thought he meant a condition passed as an argument into a function.

File: src/languages/csharp.ts (hunk 1 above)
Comment (auscompgeek) [reply src/languages/csharp.ts]: Ah, interesting. Could you add equivalent tests to the ones I requested above for while loops too?

**Comment by AndrewDant** [comment]
@pokey as we discussed at the meetup today please update vscode-parse-tree to look at a commit on master for c# tree sitter instead of version 0.19.0, in order to fix the issue where the condition field is not recognized inside of while statements.

**Comment by AndrewDant** [comment]
@pokey I think this is ready to go now. Let me know if you see any other issues.

File: src/languages/python.ts [hunk 2]
```diff
@@ -101,6 +101,7 @@ const nodeMatchers: Partial<Record<ScopeType, NodeMatcherAlternative>> = {
     argumentMatcher("parameters", "argument_list"),
     matcher(patternFinder("call.generator_expression!"), childRangeSelector())
   ),
+  callee: "call.identifier!",
```
Comment (pokey) [inline src/languages/python.ts:104]: shouldn't this one be in #664?

File: src/typings/Types.ts [hunk 3]
```diff
@@ -96,6 +96,7 @@ export type ScopeType =
   | "argumentOrParameter"
   | "anonymousFunction"
   | "attribute"
+  | "callee"
```
Comment (pokey) [inline src/typings/Types.ts:99]: shouldn't this one be in #664 ?

File: src/test/suite/fixtures/recorded/languages/csharp/changeCondition2.yml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+languageId: csharp
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/languages/csharp/changeCondition2.yml:1]: would be good to also have a for-loop test where selection is in body of loop

Comment (AndrewDant) [reply src/test/suite/fixtures/recorded/languages/csharp/changeCondition2.yml:1]: I've added two.

**Review by pokey — APPROVED** [review]
looks good!

**Pull request merged**